### PR TITLE
[inductor] Refactor and optimize allocation calls

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111402
* __->__ #111117

This splits out changes from
https://github.com/pytorch/pytorch/pull/102625 to make things easier to
review.

This diff creates a `make_allocation()` method that extracts the logic
from `make_buffer_allocation()` while allowing us to allocate non-buffer
objects. In particular, we will use this to allocate memory pools during
memory planning.

This diff also includes a small optimization -- if the desired
allocation is contiguous, then we emit a call to `empty()` instead of
`empty_strided()` with its superfluous stride argument.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/extension_backends/extension_device.cpp [hunk 1]
```diff
@@ -116,6 +116,20 @@ at::Tensor custom__copy_from(const at::Tensor& self, const at::Tensor& dst, bool
   return dst;
 }
 
+at::Tensor custom_empty_memory_format(at::IntArrayRef size,
```
Comment (int3) [inline test/inductor/extension_backends/extension_device.cpp:119]: I copied this from `test/cpp_extensions/open_registration_extension.cpp`. Not sure why we have a bunch of definitions duplicated between these two files...

**Review by jansel — APPROVED** [review]

**Comment by malfet** [comment]
@jeanschmidt nosignal meant for reverts that were due to the green signal on PR but red on trunk, for internal CI it should be `ghfirst`. Also provide some short description what did it break

**Pull request closed without merging**